File: modules/debugger/src/main/java/com/falsepattern/zigbrains/debugger/toolchain/MSVCMetadataProvider.java
package com.falsepattern.zigbrains.debugger.toolchain;

import com.falsepattern.zigbrains.ZigBundle;
import com.falsepattern.zigbrains.debugger.settings.MSVCDownloadPermission;
import com.falsepattern.zigbrains.debugger.settings.ZigDebuggerSettings;
import com.intellij.notification.Notification;
import com.intellij.notification.NotificationType;
import com.intellij.notification.Notifications;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.ui.DialogBuilder;
import com.intellij.ui.components.JBLabel;
import com.intellij.ui.components.JBPanel;
import com.intellij.util.download.DownloadableFileService;
import lombok.Cleanup;
import lombok.val;

import javax.swing.BoxLayout;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.List;
import java.util.Objects;
import java.util.Properties;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicReference;

import static com.falsepattern.zigbrains.debugger.toolchain.ZigDebuggerToolchainService.downloadPath;

class MSVCMetadataProvider {
    private Properties cached;
    private Future<Properties> downloadMSVCProps() {
        val future = new CompletableFuture<Properties>();
        ApplicationManager.getApplication().executeOnPooledThread(() -> {
            try {
                val service = DownloadableFileService.getInstance();
                val desc = service.createFileDescription("https://falsepattern.com/zigbrains/msvc.properties", "msvc.properties");
                val downloader = service.createDownloader(List.of(desc), "Debugger metadata downloading");
                val downloadDirectory = downloadPath().toFile();
                val prop = new Properties();
                val downloadResults = downloader.download(downloadDirectory);
                for (val result : downloadResults) {
                    if (Objects.equals(result.second.getDefaultFileName(), "msvc.properties")) {
                        @Cleanup val reader = new FileReader(result.first);
                        prop.load(reader);
                    }
                }
                future.complete(prop);
            } catch (Throwable t) {
                future.completeExceptionally(t);
            }
        });
        return future;
    }

    private Properties fetchBuiltinMSVCProps() {
        val prop = new Properties();
        try {
            @Cleanup val resource = ZigDebuggerToolchainService.class.getResourceAsStream("/msvc.properties");
            if (resource == null) {
                Notifications.Bus.notify(new Notification(
                        "ZigBrains.Debugger.Error",
                        ZigBundle.message("notification.title.debugger"),
                        ZigBundle.message("notification.content.debugger.metadata.fallback.fetch.failed"),
                        NotificationType.ERROR
                ));
                return prop;
            }
            val reader = new InputStreamReader(resource);
            prop.load(reader);
        } catch (IOException ex) {
            ex.printStackTrace();
            Notifications.Bus.notify(new Notification(
                    "ZigBrains.Debugger.Error",
                    ZigBundle.message("notification.title.debugger"),
                    ZigBundle.message("notification.content.debugger.metadata.fallback.parse.failed"),
                    NotificationType.ERROR
            ));
        }
        return prop;
    }


    Properties msvcProperties() {
        if (cached != null)
            return cached;
        val settings = ZigDebuggerSettings.getInstance();
        var permission = settings.msvcConsent;
        if (permission == MSVCDownloadPermission.AskMe) {
            AtomicReference<Boolean> accepted = new AtomicReference<>(false);
            ApplicationManager.getApplication().invokeAndWait(() -> {
                val dialog = new DialogBuilder();
                dialog.setTitle("Network Request Consent");
                dialog.addCancelAction().setText("Deny");
                dialog.addOkAction().setText("Allow");
                val centerPanel = new JBPanel<>();
                centerPanel.setLayout(new BoxLayout(centerPanel, BoxLayout.Y_AXIS));
                centerPanel.add(new JBLabel("ZigBrains needs to download some metadata from the internet for debugging on Windows."));
                centerPanel.add(new JBLabel("ZigBrains will use the fallback metadata shipped inside the plugin if the request is denied."));
                centerPanel.add(new JBLabel("Would you like to allow this network request?"));
                dialog.centerPanel(centerPanel);
                accepted.set(dialog.showAndGet());
            });
            permission = settings.msvcConsent = accepted.get() ? MSVCDownloadPermission.Allow : MSVCDownloadPermission.Deny;
        }
        return switch (permission) {
            //noinspection DataFlowIssue
            case AskMe, Deny -> cached = fetchBuiltinMSVCProps();
            case Allow -> {
                val future = downloadMSVCProps();
                try {
                    yield future.get(3, TimeUnit.SECONDS);
                } catch (ExecutionException | InterruptedException | TimeoutException e) {
                    e.printStackTrace();
                    Notifications.Bus.notify(new Notification(
                            "ZigBrains.Debugger.Error",
                            ZigBundle.message("notification.title.debugger"),
                            ZigBundle.message("notification.content.debugger.metadata.downloading.failed"),
                            NotificationType.ERROR
                    ));
                    settings.msvcConsent = MSVCDownloadPermission.Deny;
                    yield cached = fetchBuiltinMSVCProps();
                }
            }
        };
    }
}

File: cidr/src/main/kotlin/com/falsepattern/zigbrains/debugger/toolchain/MSVCMetadataProvider.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.debugger.toolchain

import com.falsepattern.zigbrains.debugger.ZigDebugBundle
import com.falsepattern.zigbrains.debugger.settings.MSVCDownloadPermission
import com.falsepattern.zigbrains.debugger.settings.ZigDebuggerSettings
import com.falsepattern.zigbrains.debugger.toolchain.ZigDebuggerToolchainService.Companion.downloadPath
import com.falsepattern.zigbrains.shared.coroutine.withCurrentEDTModalityContext
import com.intellij.notification.Notification
import com.intellij.notification.NotificationType
import com.intellij.openapi.ui.DialogBuilder
import com.intellij.platform.util.progress.withProgressText
import com.intellij.ui.components.JBLabel
import com.intellij.ui.components.JBPanel
import com.intellij.util.download.DownloadableFileService
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withTimeoutOrNull
import java.io.IOException
import java.util.*
import javax.swing.BoxLayout

private val mutex = Mutex()
private var cache: Properties? = null

suspend fun msvcMetadata(): Properties {
    cache?.let { return it }
    mutex.withLock {
        cache?.let { return it }
        val settings = ZigDebuggerSettings.instance
        var permission = settings.msvcConsent
        if (permission == MSVCDownloadPermission.AskMe) {
            val allowDownload = withCurrentEDTModalityContext {
                val dialog = DialogBuilder()
                dialog.setTitle(ZigDebugBundle.message("msvc.consent.title"))
                dialog.addCancelAction().setText(ZigDebugBundle.message("msvc.consent.deny"))
                dialog.addOkAction().setText(ZigDebugBundle.message("msvc.consent.allow"))
                val centerPanel = JBPanel<JBPanel<*>>()
                centerPanel.setLayout(BoxLayout(centerPanel, BoxLayout.Y_AXIS))
                val lines = ZigDebugBundle.message("msvc.consent.body").split('\n')
                for (line in lines) {
                    centerPanel.add(JBLabel(line))
                }
                dialog.centerPanel(centerPanel)
                dialog.showAndGet()
            }
            permission = if (allowDownload) MSVCDownloadPermission.Allow else MSVCDownloadPermission.Deny
            settings.msvcConsent = permission
        }
        val data = if (permission == MSVCDownloadPermission.Allow) {
            withTimeoutOrNull(3000L) {
                downloadMSVCProps()
            } ?: run {
                Notification(
                    "zigbrains",
                    ZigDebugBundle.message("notification.title.debugger"),
                    ZigDebugBundle.message("notification.content.debugger.metadata.downloading.failed"),
                    NotificationType.ERROR
                ).notify(null)
                fetchBuiltinMSVCProps()
            }
        } else {
            fetchBuiltinMSVCProps()
        }
        cache = data
        return data
    }
}

private suspend fun downloadMSVCProps(): Properties {
    return withProgressText("Downloading debugger metadata") {
        val service = DownloadableFileService.getInstance()
        val desc = service.createFileDescription("https://falsepattern.com/zigbrains/msvc.properties", "msvc.properties")
        val downloader = service.createDownloader(listOf(desc), "Debugger metadata downloading")
        val downloadDirectory = downloadPath().toFile()
        val prop = Properties()
        val downloadResults = runBlocking {
            downloader.download(downloadDirectory)
        }
        for (result in downloadResults) {
            if (result.second.defaultFileName == "msvc.properties") {
                result.first.reader().use { prop.load(it) }
            }
        }
        return@withProgressText prop
    }
}

private fun fetchBuiltinMSVCProps(): Properties {
    val prop = Properties()
    try {
        val resource = ZigDebuggerToolchainService::class.java.getResourceAsStream("/msvc.properties") ?: throw IOException("null")
        resource.reader().use { prop.load(it) }
    } catch (ex: IOException) {
        ex.printStackTrace()
        Notification(
            "zigbrains",
            ZigDebugBundle.message("notification.title.debugger"),
            ZigDebugBundle.message("notification.content.debugger.metadata.fallback.parse.failed"),
            NotificationType.ERROR
        ).notify(null)
    }
    return prop
}